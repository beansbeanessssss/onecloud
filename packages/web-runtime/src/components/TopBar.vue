<template>
  <header
    id="oc-topbar"
    class="uk-flex uk-flex-middle uk-flex-between oc-px-m"
    :aria-label="$gettext('Top bar')"
  >
    <div class="topbar-gap uk-flex uk-flex-middle uk-flex-between">
      <applications-menu v-if="applicationsList.length" :applications-list="applicationsList" />
      <router-link ref="navigationSidebarLogo" to="/">
        <oc-img :src="logoImage" :alt="sidebarLogoAlt" />
      </router-link>
    </div>
    <div class="portal-wrapper">
      <portal-target name="app.runtime.header" multiple></portal-target>
    </div>
    <div class="topbar-gap uk-flex uk-flex-middle uk-flex-between">
      <feedback-link v-if="isFeedbackLinkEnabled" />
      <notifications v-if="isNotificationBellEnabled" />
      <user-menu
        v-if="isUserMenuEnabled"
        :user-id="userId"
        :user-display-name="userDisplayName"
        :applications-list="applicationsList"
      />
    </div>
  </header>
</template>

<script>
import { mapGetters } from 'vuex'

import ApplicationsMenu from './ApplicationsMenu.vue'
import UserMenu from './UserMenu.vue'
import Notifications from './Notifications.vue'
import FeedbackLink from './FeedbackLink.vue'

export default {
  components: {
    FeedbackLink,
    Notifications,
    ApplicationsMenu,
    UserMenu
  },
  props: {
    userId: {
      type: String,
      required: false,
      default: ''
    },
    userDisplayName: {
      type: String,
      required: false,
      default: ''
    },
    applicationsList: {
      type: Array,
      required: false,
      default: () => []
    },
    activeNotifications: {
      type: [Array, Boolean],
      required: false,
      default: () => []
    }
  },
  computed: {
    ...mapGetters(['configuration', 'user']),

    sidebarLogoAlt() {
      return this.$gettext('Navigate to all files page')
    },

    logoImage() {
      return this.configuration.theme.logo.sidebar
    },

    isFeedbackLinkEnabled() {
      return !this.configuration.options.disableFeedbackLink
    },

    isNotificationBellEnabled() {
      return this.user?.id && this.activeNotifications.length
    },

    isUserMenuEnabled() {
      return this.user?.id
    }
  }
}
</script>

<style lang="scss">
#oc-topbar {
  height: 60px;
  position: sticky;
  z-index: 2;
  img {
    height: 42px;
    image-rendering: auto;
    image-rendering: crisp-edges;
    image-rendering: pixelated;
    image-rendering: -webkit-optimize-contrast;
    user-select: none;
  }
  .topbar-gap {
    gap: 30px;
  }
  @media only screen and (max-width: 960px) {
    .topbar-gap {
      gap: 0.5rem;
    }
  }
}
</style>
