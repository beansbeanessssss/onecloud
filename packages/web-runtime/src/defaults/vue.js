import Vue from 'vue'
import WebPlugin from '../plugins/web'
import Avatar from '../components/Avatar.vue'
import focusMixin from '../mixins/focusMixin'
import lifecycleMixin from '../mixins/lifecycleMixin'
import VueRouter from 'vue-router'
import Vuex from 'vuex'

Vue.use(Vuex)
Vue.use(VueRouter)
Vue.use(WebPlugin)

Vue.component('AvatarImage', Avatar)

Vue.mixin(focusMixin)
Vue.mixin(lifecycleMixin)

// externalize Vue - this is not the Vue instance but the class
window.Vue = Vue

export default Vue
