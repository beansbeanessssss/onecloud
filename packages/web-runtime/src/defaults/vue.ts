import Vue from 'vue'
import WebPlugin from '../plugins/web'
import Avatar from '../components/Avatar.vue'
import focusMixin from '../mixins/focusMixin'
import lifecycleMixin from '../mixins/lifecycleMixin'

Vue.use(WebPlugin)

Vue.component('AvatarImage', Avatar)

Vue.mixin(focusMixin)
Vue.mixin(lifecycleMixin)

// externalize Vue - this is not the Vue instance but the class
window.Vue = Vue

export default Vue
