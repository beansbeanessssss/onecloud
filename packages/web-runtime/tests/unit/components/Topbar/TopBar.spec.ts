import { computed } from 'vue'
import TopBar from 'web-runtime/src/components/Topbar/TopBar.vue'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions
} from 'web-test-helpers'

const mockUseEmbedMode = jest.fn().mockReturnValue({ isEnabled: computed(() => false) })

jest.mock('@ownclouders/web-pkg', () => ({
  ...jest.requireActual('@ownclouders/web-pkg'),
  useEmbedMode: jest.fn().mockImplementation(() => mockUseEmbedMode())
}))

describe('Top Bar component', () => {
  it('Displays applications menu', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.find('applications-menu-stub').exists()).toBeTruthy()
  })
  describe('notifications bell', () => {
    it('should display in authenticated context if announced via capabilities', () => {
      const { wrapper } = getWrapper({
        capabilities: {
          notifications: { 'ocs-endpoints': ['list', 'get', 'delete'] }
        }
      })
      expect(wrapper.find('notifications-stub').exists()).toBeTruthy()
    })
    it('should not display in an unauthenticated context', () => {
      const { wrapper } = getWrapper({
        isUserContextReady: false,
        capabilities: {
          notifications: { 'ocs-endpoints': ['list', 'get', 'delete'] }
        }
      })
      expect(wrapper.find('notifications-stub').exists()).toBeFalsy()
    })
    it('should not display if endpoint list is missing', () => {
      const { wrapper } = getWrapper({
        capabilities: { notifications: { 'ocs-endpoints': [] } }
      })
      expect(wrapper.find('notifications-stub').exists()).toBeFalsy()
    })
  })
  it.each(['applications-menu', 'theme-switcher', 'feedback-link', 'notifications', 'user-menu'])(
    'should hide %s when mode is "embed"',
    (componentName) => {
      mockUseEmbedMode.mockReturnValue({
        isEnabled: computed(() => true)
      })

      const { wrapper } = getWrapper()
      expect(wrapper.find(`${componentName}-stub`).exists()).toBeFalsy()
    }
  )
  it.each(['applications-menu', 'theme-switcher', 'feedback-link', 'notifications', 'user-menu'])(
    'should not hide %s when mode is not "embed"',
    (componentName) => {
      mockUseEmbedMode.mockReturnValue({
        isEnabled: computed(() => false)
      })

      const { wrapper } = getWrapper({
        capabilities: {
          notifications: { 'ocs-endpoints': ['list', 'get', 'delete'] }
        }
      })
      expect(wrapper.find(`${componentName}-stub`).exists()).toBeTruthy()
    }
  )
})

const getWrapper = ({ capabilities = {}, isUserContextReady = true } = {}) => {
  const mocks = { ...defaultComponentMocks() }
  const storeOptions = { ...defaultStoreMockOptions }
  storeOptions.getters.configuration.mockImplementation(() => ({
    options: { disableFeedbackLink: false }
  }))

  const store = createStore(storeOptions)
  return {
    wrapper: shallowMount(TopBar, {
      props: {
        applicationsList: ['testApp']
      },
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              authState: { userContextReady: isUserContextReady },
              capabilityState: { capabilities }
            }
          }),
          store
        ],
        stubs: { 'router-link': true, 'portal-target': true, notifications: true },
        mocks,
        provide: mocks
      }
    })
  }
}
