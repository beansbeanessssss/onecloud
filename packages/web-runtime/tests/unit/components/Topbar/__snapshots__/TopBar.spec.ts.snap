// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Bar component Displays applications menu 1`] = `
<header aria-label="Top bar" class="oc-flex oc-flex-middle oc-flex-between oc-px-m" id="oc-topbar">
  <div class="oc-topbar-left oc-flex oc-flex-middle oc-flex-between">
    <applications-menu-stub applicationslist="[object Object],[object Object],[object Object]"></applications-menu-stub>
    <router-link-stub to="/"></router-link-stub>
  </div>
  <div class="portal-wrapper">
    <portal-target-stub multiple="" name="app.runtime.header"></portal-target-stub>
  </div>
  <div class="oc-topbar-right oc-flex oc-flex-middle oc-flex-between">
    <portal-target-stub multiple="" name="app.runtime.header.right"></portal-target-stub>
  </div>
  <portal order="50" to="app.runtime.header.right">
    <theme-switcher-stub></theme-switcher-stub>
    <feedback-link-stub></feedback-link-stub>
  </portal>
  <portal order="100" to="app.runtime.header.right">
    <!--v-if-->
    <user-menu-stub applicationslist="[object Object],[object Object],[object Object]"></user-menu-stub>
  </portal>
</header>
`;
